src/orange/surf/Involute.hh

```diff
@@ -0,0 +1,347 @@
+                                  real_type a,
+                                  real_type sign,
+                                  real_type tmin,
+                                  real_type tmax)
+    : origin_(origin), r_b_(radius), a_(a), sign_(sign), tmin_(tmin), tmax_(tmax)
+{
+    double const pi = 3.14159265358979323846;
+    CELER_EXPECT(radius > 0);
+    CELER_EXPECT(a > 0);
+    CELER_EXPECT(abs(tmax) < 2 * pi + abs(tmin));
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from raw data.
+ */
+template<class R>
+CELER_FUNCTION Involute::Involute(Span<R, StorageSpan::extent> data)
+    : origin_{data[0], data[1], data[2]}
+    , r_b_{data[3]}
+    , a_{data[4]}
+    , sign_{data[5]}
+    , tmin_{data[6]}
+    , tmax_{data[7]}
+{
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Determine the sense of the position relative to this surface.
+ */
+CELER_FUNCTION SignedSense Involute::calc_sense(Real3 const& pos) const
+{
+    double const pi = 3.14159265358979323846;
+
+    /*
+     * Define tolerances.
+     */
+    real_type const tol = 1e-7;
+
+    real_type const x = pos[0] - origin_[0];
+    real_type const y = pos[1] - origin_[1];
+
+    /*
+     * Calculate distance to origin and obtain t value for disance.
+     */
+    real_type const rxy2 = ipow<2>(x) + ipow<2>(y);
+    real_type const tPoint2 = (rxy2 / ipow<2>(r_b_)) - 1;
+    real_type const tPoint = sqrt(tPoint2);
+
+    /*
+     * Check if Point is on involute.
+     */
+    real_type const angle = tPoint + a_;
+    real_type const xInv = r_b_ * (std::cos(angle) + tPoint * std::sin(angle));
+    real_type const yInv = r_b_ * (std::sin(angle) - tPoint * std::cos(angle));
+
+    if (abs(x - xInv) < tol && abs(y - yInv) < tol)
+    {
+        return SignedSense::on;
+    }
+
+    /*
+     * Check if point is in defined bounds.
+     */
+    if (abs(tPoint2) < ipow<2>(tmin_) - tol)
+    {
+        return SignedSense::outside;
+    }
+    if (abs(tPoint2) > ipow<2>(tmax_) + tol)
+    {
+        return SignedSense::outside;
+    }
+
+    /*
+     * Check if point is inside involute
+     */
+
+    // Calculate tangents
+    real_type a = rxy2;
+    real_type b;
+    real_type c;
+    real_type xa, xb;
+    real_type ya, yb, yc, yd;
+    real_type yalpha, ybeta;
+
+    b = -2 * ipow<2>(r_b_) * x;
+    c = ipow<4>(r_b_) - ipow<2>(y) * ipow<2>(r_b_);
+
+    xa = (-b + sqrt(ipow<2>(b) - 4 * a * c)) / (2 * a);
+    xb = (-b - sqrt(ipow<2>(b) - 4 * a * c)) / (2 * a);
+
+    ya = sqrt(ipow<2>(r_b_) - ipow<2>(xa));
+    yb = -sqrt(ipow<2>(r_b_) - ipow<2>(xa));
+    yc = sqrt(ipow<2>(r_b_) - ipow<2>(xb));
+    yd = -sqrt(ipow<2>(r_b_) - ipow<2>(xb));
+
+    b = -2 * ipow<2>(r_b_) * y;
+    c = ipow<4>(r_b_) - ipow<2>(x) * ipow<2>(r_b_);
+
+    yalpha = (-b + sqrt(ipow<2>(b) - 4 * a * c)) / (2 * a);
+    ybeta = (-b - sqrt(ipow<2>(b) - 4 * a * c)) / (2 * a);
```

Comment:
Is there a way to sensibly use the `detail::QuadraticSolver` for this calculation?

Reply:
`detail::QuadraticSolver` only results in real and positive solutions to the quadratic equation. However, since it is possible the point to construct the tangent to contain negatives, this method is required.